### PR TITLE
Adam op

### LINKED ISSUE #4377: Adam Operator.

### PR DESCRIPTION
**Pull request by tonyyang-svail**
fix #4377

Adam Gradient Descent algorithm.

moment1_out = beta1 * moment1 + (1 − beta1) * grad
moment2_out = beta2 * moment2 + (1 − beta2) * grad * grad
moment1_hat =  moment1_out / (1 - beta1^t)
moment2_hat =  moment2_out / (1 - beta2^t)
param_out = param - learning_rate * moment1_hat / (sqrt(moment2_hat) + epsilon)

### CODE REVIEW

**Comment by tonyyang-svail** [comment]
Passed its own unit test.

File: paddle/operators/adam_op.cc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace paddle {
+namespace operators {
+
+class AdamOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
```
Comment (jacquesqiao) [inline paddle/operators/adam_op.cc]: need to change to new InferShapeContextBase

File: paddle/operators/adam_op.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+};
+
+class AdamOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  AdamOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("param", "input parameter");
+    AddInput("grad", "input gradient");
+    AddInput("moment1", "first moment");
```
Comment (abhinavarora) [inline paddle/operators/adam_op.cc]: Also, we should be changing strings like `moment1`, `beta1` , etc to `moment_1` and `beta_1` because these are for the Python interface. In Python `_` is a separator for words in variable names.

File: paddle/operators/adam_op.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    float beta2 = ctx.Attr<float>("beta2");
+    float epsilon = ctx.Attr<float>("epsilon");
+
+    auto place = ctx.GetEigenDevice<Place>();
+    m1_o.device(place) = beta1 * m1 + (1 - beta1) * g;
+    m2_o.device(place) = beta2 * m2 + (1 - beta2) * g * g;
+
+    float beta1_to_t = std::pow(beta1, t);
+    float beta2_to_t = std::pow(beta2, t);
+    auto m1_hat = m1_o / (1 - beta1_to_t);
```
Comment (abhinavarora) [inline paddle/operators/adam_op.h]: We can increase the efficiency of this computation by changing the order of the computation. Please refer to the modified computation just before the section 2.1 of the Adam Paper (https://arxiv.org/abs/1412.6980)

File: paddle/operators/adam_op.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    AddOutput("moment1_out", "output first moment");
+    AddOutput("moment2_out", "output second moment");
+
+    AddAttr<int>("time_step", "time step");
+    AddAttr<float>("learning_rate", "learning rate of sgd");
+    AddAttr<float>("epsilon", "prevent divide by zero");
+    AddAttr<float>("beta1", "exponential decay for the first moment");
+    AddAttr<float>("beta2", "exponential decay for the second moment");
+    AddComment(R"DOC(
+
```
Comment (abhinavarora) [inline paddle/operators/adam_op.cc]: I think we should also cite the Adam Paper here.

File: paddle/operators/adam_op.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenScalar = framework::EigenScalar<T, MajorType, IndexType>;
```
Comment (abhinavarora) [inline paddle/operators/adam_op.h]: Are we using this?

**Comment by tonyyang-svail** [comment]
Reminder: should learning rate be an attribute or tensor variable?

**Comment by abhinavarora** [comment]
I discussed with @reyoung and as per his suggestion, learning rate and time steps should be inputs and not attributes. I am right now trying to figure out if we can pass a float/int as an input instead of a tensor with shape (1,)

**Comment by abhinavarora** [comment]
@tonyyang-svail  We have figured out how to pass learning rate as float to the op. The concerned changes have been made in the SGD_op. You can refer to this PR for more info https://github.com/PaddlePaddle/Paddle/pull/4568
